document-portal/document-portal-fuse.c

```diff
@@ -224,6 +223,40 @@ static int ensure_docdir_inode (XdpInode *parent,
                                 struct fuse_entry_param *e,
                                 XdpInode **inode_out);
 
+typedef struct fuse_session XdpOwnedFuseSession;
+
+static XdpOwnedFuseSession *
+own_fuse_session (void)
+{
+  XdpOwnedFuseSession *se = g_atomic_pointer_get (&session);
+  XdpOwnedFuseSession *invalid_session = (XdpOwnedFuseSession *) &session;
```

Comment:
I don't understand what this is supposed to be doing. From what I can see, `&session` will be of type `(struct fuse_session **)`, which you're casting to `(XdpOwnedFuseSession *)`, which in turn is an alias for `(struct fuse_session *)`.

This makes me worry that some code elsewhere will lead to memory errors. We've also now got a mix of atomic and non-atomic access to the variable, which is likely to cause problems.

What semantics are you trying to achieve here?

Reply:
This is just a lock-free implementation (a simpler Read Copy Update) to get the session swapping its value to an invalid pointer when it's used and setting it back when done.

So basically, while it's used we set it to an invalid pointer (unsing the well known address of the very same variable, so that it can't be confused with anything else) and we set it back once done. All the checks in which there may be concurrency look like they are using atomic math to me, though. As comparing those local variables shouldn't be problematic.

The same could be done using a temporary structure that holds a `GMutexLocker` and that we unset on autofree, but this one avoids doing further memory allocations.

ThreadSanitizer is not showing any trouble here though.

Reply:
This doesn't look lock free to me. From the look of it, `own_fuse_session` is essentially a spin lock: I couldn't see anything that resembled RCU. Did you try using a traditional mutex before rejecting it?

Also, GMutexLocker looks like it shouldn't involve any more memory allocations than what you've written: g_mutex_locker_new() doesn't do any heap allocations.

Reply:
I'm not saying that `GMutexLocker` allocates memory per sé, but I meant by wrapping the session instead.

However, using a normal mutex would be just as this:

```diff
diff --git a/document-portal/document-portal-fuse.c b/document-portal/document-portal-fuse.c
index f44bd91..4bd0038 100644
--- a/document-portal/document-portal-fuse.c
+++ b/document-portal/document-portal-fuse.c
@@ -88,6 +88,7 @@
 
 static GThread *fuse_thread = NULL;
 static struct fuse_session *session = NULL;
+G_LOCK_DEFINE (session);
 static char *mount_path = NULL;
 static pthread_t fuse_pthread = 0;
 static uid_t my_uid;
@@ -223,40 +224,6 @@ static int ensure_docdir_inode (XdpInode *parent,
                                 struct fuse_entry_param *e,
                                 XdpInode **inode_out);
 
-typedef struct fuse_session XdpOwnedFuseSession;
-
-static XdpOwnedFuseSession *
-own_fuse_session (void)
-{
-  XdpOwnedFuseSession *se = g_atomic_pointer_get (&session);
-  XdpOwnedFuseSession *invalid_session = (XdpOwnedFuseSession *) &session;
-
-  if (!se)
-    return NULL;
-
-  while (se == invalid_session)
-    se = g_atomic_pointer_get (&session);
-
-  while (!g_atomic_pointer_compare_and_exchange (&session, se, invalid_session))
-    ;
-
-  return se;
-}
-
-static void
-release_fuse_session (XdpOwnedFuseSession *se)
-{
-  gpointer invalid_session = (gpointer) &session;
-
-  g_assert (se != NULL);
-  g_assert (se != invalid_session);
-  g_assert (g_atomic_pointer_get (&session) == invalid_session);
-
-  g_atomic_pointer_set (&session, se);
-}
-
-G_DEFINE_AUTOPTR_CLEANUP_FUNC (XdpOwnedFuseSession, release_fuse_session)
-
 static gboolean
 app_can_write_doc (PermissionDbEntry *entry, const char *app_id)
 {
@@ -1720,16 +1687,15 @@ static gboolean
 invalidate_doc_domain (gpointer user_data)
 {
   g_autoptr(XdpDomain) doc_domain = user_data;
-  g_autoptr(XdpOwnedFuseSession) se = NULL;
   ino_t parent_ino;
 
   g_atomic_int_set (&doc_domain->doc_queued_invalidate, 0);
 
   parent_ino = xdp_inode_to_ino (doc_domain->parent_inode);
-  se = own_fuse_session ();
 
-  if (se && g_atomic_int_get (&doc_domain->parent_inode->kernel_ref_count) > 0)
-    fuse_lowlevel_notify_inval_entry (se, parent_ino, doc_domain->doc_id,
+  XDP_AUTOLOCK (session);
+  if (session && g_atomic_int_get (&doc_domain->parent_inode->kernel_ref_count) > 0)
+    fuse_lowlevel_notify_inval_entry (session, parent_ino, doc_domain->doc_id,
                                       strlen (doc_domain->doc_id));
 
   return FALSE;
@@ -3215,9 +3181,9 @@ xdp_fuse_thread (gpointer data)
   g_atomic_pointer_set (&session, se);
   xdp_fuse_mainloop (se, &loop_config);
 
-  se = own_fuse_session ();
+  XDP_AUTOLOCK (session);
   fuse_session_unmount (se);
-  g_atomic_pointer_set (&session, NULL);
+  session = NULL;
   fuse_session_destroy (se);
 
   return NULL;
@@ -3322,13 +3288,9 @@ xdp_fuse_init (GError **error)
 void
 xdp_fuse_exit (void)
 {
-  XdpOwnedFuseSession *se = own_fuse_session ();
+  XDP_AUTOLOCK (session);
 
-  if (se)
-    {
-      fuse_session_exit (se);
-      g_clear_pointer (&se, release_fuse_session);
-    }
+  g_clear_pointer (&session, fuse_session_exit);
 
   if (fuse_pthread)
     pthread_kill (fuse_pthread, SIGHUP);
@@ -3381,13 +3343,12 @@ xdp_fuse_invalidate_doc_app (const char *doc_id,
                              const char *opt_app_id)
 {
   g_autoptr(GArray) invalidates = NULL;
-  g_autoptr(XdpOwnedFuseSession) se = NULL;
+  XDP_AUTOLOCK (session);
   int i;
 
-  se = own_fuse_session ();
   /* This can happen if fuse is not initialized yet for the very
      first dbus message that activated the service */
-  if (se == NULL)
+  if (session == NULL)
     return;
 
   g_debug ("invalidate %s/%s", doc_id, opt_app_id ? opt_app_id : "*");
@@ -3420,12 +3381,12 @@ xdp_fuse_invalidate_doc_app (const char *doc_id,
 
       if (invalidate->filename)
         {
-          fuse_lowlevel_notify_inval_entry (se, invalidate->ino,
+          fuse_lowlevel_notify_inval_entry (session, invalidate->ino,
                                             invalidate->filename, strlen (invalidate->filename));
           g_free (invalidate->filename);
         }
       else
-        fuse_lowlevel_notify_inval_inode (se, invalidate->ino, 0, 0);
+        fuse_lowlevel_notify_inval_inode (session, invalidate->ino, 0, 0);
     }
 }
 
```

Would you prefer this better?

Reply:
Ok, I've done this change.

However the side of the diff that this comment was pointing to is in a commit that I've still in the branch (could be changed), but ideally trying to address a safer access to the session while doing operations, avoiding that the fuse thread would just drop it earlier.

And yeah, could probably use mutexes, but it would make things worse imho.